### ISSUE TITLE
Using identifiers reserved for the implementation

### ISSUE THREAD

**Issue by ianmccul** [Cytnx-dev/Cytnx#460]
I see that Cytnx uses a lot of identifiers of the form underscore followed by a capital letter.  Eg all of the header file include guards `_H....`, and also as other identifiers (function names, variables etc).

This isn't a good idea: identifiers starting with an underscore then a capital letter are reserved for the implementation and should not appear in user code. Eg, the standard library might use them as global variables or macros. This actually happened once, I was debugging some student's code that was producing some weird error and we couldn't figure out what was going on, until I noticed that one of the variables was `_P`, and it turned out that that version of gcc had defined it as a macro. Note that the implementation of the standard library must **always** use such reserved identifiers in the implementation, because there is no way for the implementation to guard against users `#define` etc for non-reserved identifiers.

From https://en.cppreference.com/w/cpp/language/identifiers :

The following forms are reserved:

- in the global namespace, identifiers that begin with an underscore
- identifiers that contain a double underscore or begin with an underscore followed by an uppercase letter, except the following identifiers: [follows short list of standard identifiers that would otherwise be in the reserved list]

“Reserved” here means that the standard library headers <span class="mw-geshi cpp source-cpp"><span class="co2">#define</span></span>
 or declare such identifiers for their internal needs, the compiler may 
predefine non-standard identifiers of that kind, and that name mangling 
algorithm may assume that some of these identifiers are not in use. If 
the programmer uses such identifiers, the program is ill-formed, no 
diagnostic required.

In addition, it is undefined behavior to <span class="mw-geshi cpp source-cpp"><span class="co2">#define</span></span> or <span class="mw-geshi cpp source-cpp"><span class="co2">#undef</span></span> certain names in a translation unit, see <a href="https://en.cppreference.com/w/cpp/preprocessor/replace" title="cpp/preprocessor/replace">reserved macro names</a> for more details.

Basically, avoid any identifier that starts with an underscore. If you really want to start with an underscore, it must be followed by a lowercase letter or a digit, and it should be in some namespace/class etc and not in the global namespace.

**Comment by kaihsin** [comment]
Yes. In fact. The best way to deal with this guard is just use `#pragma once`

**Comment by kaihsin** [comment]
We need to keep an eye on this when doing refactor. But for now I would'nt bother much about this.

**Comment by ianmccul** [comment]
> We need to keep an eye on this when doing refactor. But for now I would'nt bother much about this.

Yes I agree, it obviously isn't causing problems now and the chances of a problem in the future are small, BUT it might lead to some crazy hard-to-diagnose problem, so it is good to avoid such identifiers in new code, and take the opportunity to remove existing uses when refactoring code.

**Comment by yingjerkao** [comment]
Fixed in #527

**Comment by ianmccul** [comment]
Reopening because the codebase still contains a large number of identifiers that are reserved. 11,660, if I got my regexp correct.

_Rin: 4,431 occurrences
_Lin: 4,399 occurrences

Some others:
```text
_TNs
_TNinB_REDUCE_
_Save
_Load
_Init_byptr
__static_random_device
__SII
__ScII
_A
_Ins
_Tn
_T_init
_Arnoldi_internal_complex
_Arnoldi_internal_real
_Lanczos_internal
_Arnoldi
_Lanczos
_Tensordot_generic
_Tl
_Tr
__blasINTsize__
__ARPACK_WRAPPER_H_INCLUDED__
__device
__ntwk
__nttype
__pybind_device
```

**Comment by pcchen** [comment]
## Status: the two concrete hazards are fixed; remaining scope is an audit + naming-policy question

Progress from the merged 2026-07 work:

- **#983** removed `#define __PRETTY_FUNCTION__ __FUNCTION__` (defining a reserved identifier on MSVC) — error reports now use the project-owned `CYTNX_FUNC_NAME` macro instead.
- **#981** removed the C `complex.h` macro-`I` collision from all pybind TUs and renamed the colliding `template <std::size_t I>` parameter to `Idx` in `include/Type.hpp`, with an include-order canary test guarding the public headers.

Still in scope for this issue:

- A residual audit of genuinely reserved spellings (double underscore / underscore+capital anywhere; single leading underscore at global scope) — e.g. `__blasINTsize__` is still exposed, and the `__ScII` init-interface global goes away with the pending #1011.
- The broader *style* question (single leading underscore on members/helpers is legal but discouraged) is being consolidated in the naming-convention discussion (@IvanaGyro's v2.0.0 unification proposal in the #1001/#980 threads; a "Naming" section for the agent guide was also suggested on #1008) — suggest this issue stay narrowly about the *reserved* (UB) spellings and defer style to that effort.

*Posted by Claude Code on behalf of @pcchen*

**Comment by IvanaGyro** [comment]
Can we close this?

**Comment by ianmccul** [comment]
I suggest updating the AI guidance, it is a bit vague on this. `CLAUDE.md` says "No leading underscore in new code (strict) ... new members/locals must not start with _."  But it doesn't say anything about removing existing reserved identifiers, in fact it discourages unrelated naming churn. It also fails to distinguish the actual C++ rules:
- `__name` (two underscores together anywhere in an identifier): reserved everywhere
- `_Uppercase`: reserved everwhere
- `_lowercase`: reserved only at global scope - OK in class scope or namespace
